### PR TITLE
BUG: Handle subarrays in descr_to_dtype

### LINKED ISSUE #13431: np.load() "invalid shape in fixed-type tuple" in NumPy 1.16.0
In NumPy 1.16.0, `np.load()` fails to load some files created by `np.save()`, saying "ValueError: invalid shape in fixed-type tuple."

The problem seems to be that #12358 changed `np.lib.format._read_array_header()`.  It used to pass the descr to `np.dtype()`, but now it uses a new function `descr_to_dtype()` which fails for some valid inputs.

We have significant amounts of data stored on disk with formats which NumPy 1.16.0 cannot load, so this is a showstopper for upgrading beyond 1.15.

### Reproducing code example:
This fails in NumPy 1.16.0, but works in all prior versions:
```python
import numpy as np
a = np.zeros(1, np.dtype([('c', ('<f8', (5,)), (2,))]))
np.save('a.npy', a)
np.load('a.npy')
```

### Error message:
```
Traceback (most recent call last):
  File "numpy/lib/npyio.py", line 440, in load
    pickle_kwargs=pickle_kwargs)
  File "numpy/lib/format.py", line 678, in read_array
    shape, fortran_order, dtype = _read_array_header(fp, version)
  File "numpy/lib/format.py", line 563, in _read_array_header
    dtype = descr_to_dtype(d['descr'])
  File "numpy/lib/format.py", line 282, in descr_to_dtype
    dt = numpy.dtype((descr_to_dtype(descr_str), shape))
  File "numpy/lib/format.py", line 282, in descr_to_dtype
    dt = numpy.dtype((descr_to_dtype(descr_str), shape))
ValueError: invalid shape in fixed-type tuple.
```

### Numpy/Python version information:
NumPy 1.16.0
Python 3.6.6

### PR DESCRIPTION
**Pull request by mattip**
Fixes #13431.

There are alternative spellings of `dtype=[('c', '<f8', (2, 5))]`, handle the `dtype=[('c', ('<f8', (5,)), (2,))]` variant.

### CODE REVIEW

**Comment by charris** [comment]
The test failure was the matmul heisenbug.

**Review by seberg — APPROVED** [review]
Strange beast, nested subfields... But they do seem to work fine for all things (including being resolved at arbitrary depth when there are no fields left).

Anyway, LGTM, will merge soon.

File: numpy/lib/format.py [hunk 1]
```diff
@@ -261,12 +261,13 @@ def dtype_to_descr(dtype):
...
-    empty fields with name,format == '', '|Vn' are added as padding.
+    (name, format, [shape]) tuples where format may be a str or a tuple.
+    Offsets are not explicitly saved, rather empty fields with
+    name, format == '', '|Vn' are added as padding.
 
     This function reverses the process, eliminating the empty padding fields.
     '''
-    if isinstance(descr, (str, dict)):
+    if isinstance(descr, (str, dict, tuple)):
         # No padding removal needed
```
Comment (eric-wieser) [inline numpy/lib/format.py]: What happens if this is a subarray of structured types?

Comment (eric-wieser) [reply numpy/lib/format.py]:
```
s = np.dtype([('a', np.int8), ('b', np.int16), ('c', np.int32)], align=True)
s_sub = np.dtype((s, (3,)))
```

Comment (eric-wieser) [reply numpy/lib/format.py]: I think you need to recurse for subarray types

Comment (seberg) [reply numpy/lib/format.py]: That is an interesting case. Top level subarrays are degenerated on arrays (they are added to the dimensions of the array), cannot quickly find a way to create an array with such a dtype, but it somewhat feels like there may have been strange ways to do it.

Comment (seberg) [reply numpy/lib/format.py]:
```
s = np.dtype([('a', np.int8), ('b', np.int16), ('c', np.int32)], align=True)
s_sub = np.dtype((s, (1,1)))
arr = np.zeros(3, s_sub)
print(arr.shape, arr.dtype)
arr = np.ndarray(shape=3, buffer=arr, dtype=s_sub)
print(arr.shape, arr.dtype)
```

Comment (eric-wieser) [reply numpy/lib/format.py]: Also, watch out for structured types like `(int, [('fields', int)])` which have a non-void base

Comment (seberg) [reply numpy/lib/format.py]: Yeah, this one is still broken (although maybe the original issue is solved and this is just another issue). Had a too shallow look at this probably, though :/.

Comment (eric-wieser) [reply numpy/lib/format.py]: No need for the subarray to be at the top level to hit this code-path - nest it inside a structured one.

**Review by eric-wieser — COMMENTED** [review]
Trying to unstick my pending comment...

File: numpy/lib/format.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/lib/format.py]: No need for the subarray to be at the top level to hit this code-path - nest it inside a structured one.

**Comment by eric-wieser** [comment]
Here's the case that this handles incorrectly:
```
dt = np.dtype([
    ('x', np.dtype((
        np.dtype({'names':['a','b'], 'formats':['i1','i1'], 'offsets':[0,4], 'itemsize':8}),
        (3,)
    )))
])
assert descr_to_dtype(dt.descr) == dt
```

**Comment by mattip** [comment]
I think we should fail to create a dtype with `(int, [('fields', int)])`. It does not seem to fit into any of the categories of dtypes we [should parse](http://www.numpy.org/devdocs/reference/arrays.dtypes.html#specifying-and-constructing-data-types). 

In any case, its `descr` attribute does not provide the information to reconstruct it, so `lib.format.dtype_to_descr` fails. If needed, let's handle that in a different issue/PR

**Comment by eric-wieser** [comment]
Agreed that the non-void strict is not important. We should still support arbitrarily nested subarrays though.

**Comment by mattip** [comment]
I think the last commit fixed parsing nested subarrays, at least the tests with the new dtypes pass.

File: numpy/lib/tests/test_format.py [hunk 2]
```diff
@@ -628,6 +629,24 @@ def test_pickle_disallow():
...
+@pytest.mark.parametrize('dt', [
+    np.dtype(np.dtype([('a', np.int8),
+                       ('b', np.int16),
+                       ('c', np.int32),
+                      ], align=True),
+             (3,)),
+    np.dtype([('x', ([('a', '|i1'),
+                      ('', '|V3'),
+                      ('b', '|i1'),
+                      ('', '|V3'),
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply numpy/lib/tests/test_format.py]: To hit the failing path, you need to use ` np.dtype({'names':['a','b'], 'formats':['i1','i1'], 'offsets':[0,4], 'itemsize':8})` as the inner type here, not `[('a', '|i1'), ('', '|V3'), ('b', '|i1'), ('', '|V3')]`

[the remark this answers is not in the document]
Comment (mattip) [reply numpy/lib/tests/test_format.py]: that passes too

File: numpy/lib/tests/test_format.py [hunk 3]
```diff
@@ -628,6 +629,36 @@ def test_pickle_disallow():
...
+                     ],
+                     (3,)),
+               (4,),
+             )]),
+    np.dtype([('x', np.dtype({'names':['a','b'],
+                              'formats':['i1','i1'],
+                              'offsets':[0,4],
+                              'itemsize':8,
+                             },
+                    (3,)),
```
Comment (eric-wieser) [inline numpy/lib/tests/test_format.py]: I don't think `dtype(dict, tuple)` is legal, which will cause an error during test collection

File: numpy/lib/tests/test_format.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eric-wieser) [reply numpy/lib/tests/test_format.py]: The full type I use in a comment above still fails

File: numpy/lib/tests/test_format.py (hunk 3 above)
Comment (mattip) [reply numpy/lib/tests/test_format.py]: tests are passing

File: numpy/lib/tests/test_format.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/lib/tests/test_format.py]: Could you provide a complete example of a dtype that fails to roundtrip?

[the remark this answers is not in the document]
Comment (eric-wieser) [reply numpy/lib/tests/test_format.py]: [yes](https://github.com/numpy/numpy/pull/13433#issuecomment-488568995)

File: numpy/lib/tests/test_format.py (hunk 3 above)
Comment (eric-wieser) [reply numpy/lib/tests/test_format.py]:
```
>>> np.dtype(int, "this argument is ignored")
dtype('int32')
```

This test is ignoring the `(3,)` silently, which is a different bug.

File: numpy/lib/tests/test_format.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/lib/tests/test_format.py]: test added and fixed

**Comment by mattip** [comment]
> I don't think the `dict` path can ever be hit

It seems not, removing

**Comment by seberg** [comment]
This stuff still confuses me a bit, but it does seem the test should cover the interesting corner cases, so can probably merge.

File: numpy/lib/tests/test_format.py [hunk 4]
```diff
@@ -628,6 +629,57 @@ def test_pickle_disallow():
...
+            )),
+        (4,)
+        )))
+        ]),
+    np.dtype([
+        ('a', np.dtype((
+            np.dtype((
+                np.dtype((
+                    np.dtype([
+                        ('a', int)
```
Comment (seberg) [inline numpy/lib/tests/test_format.py]:
```suggestion
                        ('a', int),
                        ('b', np.dtype({'names':['a','b'],
                                        'formats':['i1','i1'],
                                        'offsets':[0,4],
                                        'itemsize':8})),
```

Finally, this is what will make things fail...

File: numpy/lib/format.py [hunk 5]
```diff
@@ -261,15 +261,21 @@ def dtype_to_descr(dtype):
...
+    if isinstance(descr, str):
         # No padding removal needed
         return numpy.dtype(descr)
-
+    elif isinstance(descr, tuple):
+        if isinstance(descr[0], list):
+            # subtype, will always have a shape descr[1]
+            dt = descr_to_dtype(descr[0])
+            return numpy.dtype((dt, descr[1]))
+        return numpy.dtype(descr)
```
Comment (seberg) [inline numpy/lib/format.py]:
```suggestion
        assert len(descr) == 2
        return descr_to_dtype(descr[0]), descr[1]
```

Is that assert correct here, since it is not a list around it, there cannot be a field name, so it must have two entries, right?

Comment (eric-wieser) [reply numpy/lib/format.py]: Almost looks correct, but this should be `np.dtype((descr_to_dtype(descr[0]), descr[1]))`

File: numpy/lib/format.py [hunk 6]
```diff
@@ -261,15 +261,21 @@ def dtype_to_descr(dtype):
...
 
     This function reverses the process, eliminating the empty padding fields.
     '''
-    if isinstance(descr, (str, dict)):
+    if isinstance(descr, str):
         # No padding removal needed
         return numpy.dtype(descr)
-
+    elif isinstance(descr, tuple):
+        if isinstance(descr[0], list):
```
Comment (seberg) [inline numpy/lib/format.py]: You can just remove this `if` and ignore my other comment.

Comment (mattip) [reply numpy/lib/format.py]: adopted

**Comment by charris** [comment]
close/reopen

**Comment by seberg** [comment]
Ok, putting this in then. What I am not quite sure is whether there is some issue that should be opened here, may come back to it, but it will be a fringe issue in any case, I suppose.
